src/plugins/data/public/data_sources/datasource_selector/datasource_selectable.tsx

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import React, { useEffect } from 'react';
+import { DataSourceSelector } from './datasource_selector';
+import { DataSourceType } from '../datasource_services';
+import { DataSourceGroup, DataSourceOption, DataSourceSelectableProps } from './types';
+import { ISourceDataSet } from '../datasource/types';
+import { IndexPattern } from '../../index_patterns';
+
+// Mapping between datasource type and its display name.
+const DATASOURCE_TYPE_DISPLAY_NAME_MAP = {
+  DEFAULT_INDEX_PATTERNS: 'Index patterns',
+  s3glue: 'Amazon S3',
+  spark: 'Spark',
+};
+
+type DataSetType = string | IndexPattern;
+
+export const DataSourceSelectable = ({
+  dataSources, // list of all available datasource connections.
+  dataSourceOptionList, // combo box renderable option list derived from dataSources
```

Comment:
Some use cases out there already are somewhat they want additional operations on the data when general optionList is done preparation, like filtering, pre/post pending data, manipulating option, maybe moving all these into the component can work, but I just feel it requires more works to make it support all these functionalities while keep it generic.

Reply:
The issue is that it then becomes a part of the API and changing it is a breaking change. To avoid that I wanted to only keep the props here that are absolutely needed. If we need that level of control we can add it later. Right now im more worried that keeping it this flexible actually puts us at the risk of having different uses of this component behave differently. What we want is for all instances of this component to render somewhat similarly. 

Can we move this into the Selectable component for now and pull it out later if needed?

Reply:
Will address after this PR merge, Tracking [issue](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5211).